File: packages/awiki_im_core/lib/src/models/local_state_upgrade.dart
enum LocalStateUpgradeEligibility { notRequired, required }

class LocalStateUpgradeInspection {
  const LocalStateUpgradeInspection({
    required this.eligibility,
    required this.sourceSchemaVersion,
    required this.targetSchemaVersion,
  });

  final LocalStateUpgradeEligibility eligibility;
  final int sourceSchemaVersion;
  final int targetSchemaVersion;
}

enum LocalStateUpgradeStatus { notRequired, completed }

class LocalStateUpgradeResult {
  const LocalStateUpgradeResult({
    required this.status,
    required this.sourceSchemaVersion,
    required this.targetSchemaVersion,
    required this.migratedPersonas,
    required this.migratedConversations,
    required this.unresolvedMessages,
    required this.aliasCount,
    required this.backupAvailable,
    this.aliasMappings = const <LocalStateConversationAliasMapping>[],
  });

  final LocalStateUpgradeStatus status;
  final int sourceSchemaVersion;
  final int targetSchemaVersion;
  final int migratedPersonas;
  final int migratedConversations;
  final int unresolvedMessages;
  final int aliasCount;
  final bool backupAvailable;
  final List<LocalStateConversationAliasMapping> aliasMappings;
}

class LocalStateConversationAliasMapping {
  const LocalStateConversationAliasMapping({
    required this.ownerIdentityId,
    required this.ownerDid,
    required this.legacyConversationId,
    required this.canonicalConversationId,
  });

  final String ownerIdentityId;
  final String ownerDid;
  final String legacyConversationId;
  final String canonicalConversationId;
}

class LocalStateRestoreResult {
  const LocalStateRestoreResult({
    required this.restoredSchemaVersion,
    required this.targetSafetyCopyAvailable,
  });

  final int restoredSchemaVersion;
  final bool targetSafetyCopyAvailable;
}

File: crates/im-core-dart/src/dto/local_state_upgrade.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DartLocalStateUpgradeEligibility {
    NotRequired,
    Required,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DartLocalStateUpgradeInspection {
    pub eligibility: DartLocalStateUpgradeEligibility,
    pub source_schema_version: i64,
    pub target_schema_version: i64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DartLocalStateUpgradeStatus {
    NotRequired,
    Completed,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DartLocalStateUpgradeResult {
    pub status: DartLocalStateUpgradeStatus,
    pub source_schema_version: i64,
    pub target_schema_version: i64,
    pub migrated_personas: u64,
    pub migrated_conversations: u64,
    pub unresolved_messages: u64,
    pub alias_count: u64,
    pub backup_available: bool,
    pub alias_mappings: Vec<DartLocalStateConversationAliasMapping>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DartLocalStateConversationAliasMapping {
    pub owner_identity_id: String,
    pub owner_did: String,
    pub legacy_conversation_id: String,
    pub canonical_conversation_id: String,
}

impl From<im_core::LocalStateUpgradeInspection> for DartLocalStateUpgradeInspection {
    fn from(value: im_core::LocalStateUpgradeInspection) -> Self {
        Self {
            eligibility: match value.eligibility {
                im_core::LocalStateUpgradeEligibility::NotRequired => {
                    DartLocalStateUpgradeEligibility::NotRequired
                }
                im_core::LocalStateUpgradeEligibility::Required => {
                    DartLocalStateUpgradeEligibility::Required
                }
            },
            source_schema_version: value.source_schema_version,
            target_schema_version: value.target_schema_version,
        }
    }
}

impl From<im_core::LocalStateUpgradeResult> for DartLocalStateUpgradeResult {
    fn from(value: im_core::LocalStateUpgradeResult) -> Self {
        Self {
            status: match value.status {
                im_core::LocalStateUpgradeStatus::NotRequired => {
                    DartLocalStateUpgradeStatus::NotRequired
                }
                im_core::LocalStateUpgradeStatus::Completed => {
                    DartLocalStateUpgradeStatus::Completed
                }
            },
            source_schema_version: value.source_schema_version,
            target_schema_version: value.target_schema_version,
            migrated_personas: value.migrated_personas,
            migrated_conversations: value.migrated_conversations,
            unresolved_messages: value.unresolved_messages,
            alias_count: value.alias_count,
            backup_available: value.backup_available,
            alias_mappings: value
                .alias_mappings
                .into_iter()
                .map(|mapping| DartLocalStateConversationAliasMapping {
                    owner_identity_id: mapping.owner_identity_id,
                    owner_did: mapping.owner_did,
                    legacy_conversation_id: mapping.legacy_conversation_id,
                    canonical_conversation_id: mapping.canonical_conversation_id,
                })
                .collect(),
        }
    }
}
